### PR TITLE
Replace CSharpIRToken with RazorIRToken

### PR DESCRIPTION
**Pull request by rynowak**
Deletes CSharpIRToken to use the more general RazorIRToken class.

Rather than using the visitor to visit tokens, now writing a
CSharpExpresionIRNode is an 'atom', and will write its tokens itself.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor.Evolution/CodeGeneration/DesignTimeCSharpRenderer.cs [hunk 1]
```diff
@@ -278,9 +276,15 @@ public override void VisitDeclareTagHelperFields(DeclareTagHelperFieldsIRNode no
...
-            var lineMapping = new LineMapping(sourceLocation, generatedLocation);
+            if (node.Source == null)
+            {
+                return;
+            }
+
+            var source = node.Source.Value;
+            
+            var generatedLocation = new SourceSpan(Context.Writer.GetCurrentSourceLocation(), source.Length);
+            var lineMapping = new LineMapping(source, generatedLocation);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Evolution/CodeGeneration/DesignTimeCSharpRenderer.cs]: The callers don't consistently check whether or not they have a `.source` which would make this throw if handed a user-synthesized token

File: src/Microsoft.AspNetCore.Razor.Evolution/CodeGeneration/DesignTimeCSharpRenderer.cs [hunk 2]
```diff
@@ -38,14 +33,17 @@ public override void VisitCSharpExpression(CSharpExpressionIRNode node)
...
+                        if (token != null && token.IsCSharp)
                         {
-                            AddLineMappingFor(childNode);
+                            AddLineMappingFor(token);
+                            Context.Writer.Write(token.Content);
+                        }
+                        else
+                        {
+                            // There may be something else inside the expression like a Template or another extension node.
+                            Visit(node.Children[i]);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Evolution/CodeGeneration/DesignTimeCSharpRenderer.cs]: This looks like our opportunity to tighten up the set of allowed/valid nodes. For instance a template is implemented as an extension node, but it's expected to appear in an expession context.

File: src/Microsoft.AspNetCore.Razor.Evolution/CodeGeneration/DesignTimeCSharpRenderer.cs (hunk 1 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Evolution/CodeGeneration/DesignTimeCSharpRenderer.cs]: Is that a new behavior?

File: src/Microsoft.AspNetCore.Razor.Evolution/CodeGeneration/DesignTimeCSharpRenderer.cs [hunk 3]
```diff
@@ -305,14 +309,14 @@ private void AddLineMappingFor(RazorIRNode node)
 
                 Context.Writer.Write(((HtmlContentIRNode)node).Content);
             }
-            else if (node is CSharpTokenIRNode)
+            else if (node is RazorIRToken token && token.IsCSharp)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Evolution/CodeGeneration/DesignTimeCSharpRenderer.cs]: 👏

File: src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorIRLoweringPhase.cs [hunk 4]
```diff
@@ -376,11 +376,19 @@ public override void VisitExpressionSpan(ExpressionChunkGenerator chunkGenerator
                     }
                 }
 
-                _builder.Add(new CSharpTokenIRNode()
+                if (_builder.Current is CSharpExpressionIRNode)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorIRLoweringPhase.cs]: Why the new check?

File: src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorIRLoweringPhase.cs [hunk 5]
```diff
@@ -376,11 +376,19 @@ public override void VisitExpressionSpan(ExpressionChunkGenerator chunkGenerator
...
+                    _builder.Add(new RazorIRToken()
+                    {
+                        Content = span.Content,
+                        Kind = RazorIRToken.TokenKind.CSharp,
+                        Source = BuildSourceSpanFromNode(span),
+                    });
+                }
+                else
+                {
+                    throw null;
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorIRLoweringPhase.cs]: O_o

File: test/Microsoft.AspNetCore.Razor.Evolution.Test/Intermediate/RazorIRAssert.cs [hunk 6]
```diff
@@ -203,7 +203,10 @@ public static void CSharpExpression(string expected, RazorIRNode node)
                 var content = new StringBuilder();
                 for (var i = 0; i < cSharp.Children.Count; i++)
                 {
-                    content.Append(((CSharpTokenIRNode)cSharp.Children[i]).Content);
+                    if (cSharp.Children[i] is RazorIRToken token && token.IsCSharp)
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Razor.Evolution.Test/Intermediate/RazorIRAssert.cs]: We casted before, lets just cast now and Assert.True(token.IsCSharp) so we know when we make a behavior change in the future.

File: src/Microsoft.AspNetCore.Razor.Evolution/CodeGeneration/DesignTimeCSharpRenderer.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Evolution/CodeGeneration/DesignTimeCSharpRenderer.cs]: No, but users could do things that will make this code throw a nullref.

File: src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorIRLoweringPhase.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorIRLoweringPhase.cs]: This was just some debugging code I left in.

File: test/Microsoft.AspNetCore.Razor.Evolution.Test/Intermediate/RazorIRAssert.cs (hunk 6 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Razor.Evolution.Test/Intermediate/RazorIRAssert.cs]: Yeah this should have an `Assert.IsType` instead.

**Comment by rynowak** [comment]
:shipit:

**Comment by rynowak** [comment]
1ae0b216306e72c128c91b116775e6fa3a88a8ea

**Pull request closed without merging**